keps/sig-node/xxx-dynamic-resource-allocation/README.md

```diff
@@ -0,0 +1,1408 @@
+This might be a good place to talk about core concepts and how they relate.
+-->
+
+Scheduling is likely to be slower when many Pods request the new
+resource types, both because scheduling such a Pod involves more
+round-trips through the API server for ResourceClaimStatus updates and
+because scheduling one Pod may affect other Pods in ways that cannot
+be anticipated by the kube-scheduler. When many Pods compete for
+limited resources, multiple attempts may be needed before a suitable
+node is found.
+
+The hardware that is expected to need this more flexible allocation
+approach is going to be used by only a small subset of the pods in the
+cluster and those pods are likely to run for extended periods of time,
+so this is not a major concern.
+
+### Risks and Mitigations
+
+<!--
+What are the risks of this proposal, and how do we mitigate? Think broadly.
+For example, consider both security and how this will impact the larger
+Kubernetes ecosystem.
+
+How will security be reviewed, and by whom?
+
+How will UX be reviewed, and by whom?
+
+Consider including folks who also work outside the SIG or subproject.
+-->
+
+## Design Details
+
+<!--
+This section should contain enough information that the specifics of your
+change are understandable. This may include API specs (though not always
+required) or even code snippets. If there's any ambiguity about HOW your
+proposal will be implemented, this is the place to discuss them.
+-->
+
+### Implementation
+
+![components](./components.svg)
+
+Several components must be implemented or modified in Kubernetes:
+- The new API must be added to kube-apiserver.
+- A new controller in kube-controller-manager which generates
+  ResourceClaims from Pod ResourceClaimTemplates, similar to
+  https://github.com/kubernetes/kubernetes/tree/master/pkg/controller/volume/ephemeral
+- A kube-scheduler plugin must detect Pods which reference a
+  ResourceClaim (directly or through a template) and ensure that the
+  resource is allocated before the Pod gets scheduled, similar to
+  https://github.com/kubernetes/kubernetes/blob/master/pkg/controller/volume/scheduling/scheduler_binder.go
+- Kubelet must be extended to retrieve information from ResourceClaims
+  and then invoke local resource plugin methods. It must pass information about
+  the additional resources to the container runtime. It must detect
+  whether the container runtime has the necessary support and
+  advertise that in the cluster via node labels to simplify deployment
+  of plugins and Pod scheduling.
+
+For a resource plugin the following components are needed:
+- Some utility library similar to
+  https://github.com/kubernetes-sigs/sig-storage-lib-external-provisioner
+  and the code in driver-registrar.
+- *Resource controller*: a central component which handles resource allocation
+  by watching and modifying ResourceClaims.
+- *Resource node plugin*: a component which cooperates with kubelet to prepare
+  the usage of the resource on a node.
```

Comment:
I want to make sure I understand the proposed flow here. Let me see if I can put it into my own word. Please correct anything I got wrong.

First -- it sounds like plugins no longer "pre-advertise" their resources in the same way anymore. I.e. the api-server / scheduler no longer has a "count" of how many of a given resource a plugin is in charge of. Instead, the plugins now advertise their resources via a custom CRD, so users can still see what's available and request them in their pod spec (I assume with some sort of accounting of how many of each type are available still somewhere).

With this in place, a container within a pod spec can now "declare" what resources it would like to get access to, and a set of `ResourceClaim`s will get generated on its behalf.  Plugins on each node are then responsible for monitoring these `ResourceClaim`s and if a plugin can satisfy it, it declares that it can by adding its node's name to the `PotentialNodes` list in the `ResourceClaimStatus` field of the `ResourceClaim` (essentially locking these resources on each node until the scheduler has made a decision on which one to use to satisfy the claim).

Once this is done, the scheduler plugin kicks in to figure out which node is best suited to satisfy all of the the resource requests across all containers of the pod and eventually schedules the pod there (unlocking the resources on all nodes except the one where the pod is being scheduled).

After being scheduled, the kublet then does whatever is necessary to actually run the pod and grant it the requested resources. It does this by coordinating with the plugin to figure out what fields need to be added to the CRI / CDI interfaces to satisfy the request.